cf-keycrypt/cf-keycrypt.c

```diff
@@ -0,0 +1,388 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+
+*/
+
+#include <config.h>
+#include <cf3.defs.h>
+#include <cf3.extern.h>
+#include <openssl/err.h>
+
+#include <lastseen.h>
+#include <conversion.h>
+#include <files_hashes.h>
+#include <locks.h>
+#include <item_lib.h>
+#include <known_dirs.h>
+#include <dbm_api.h>
+/*
+#ifdef LMDB
+#include <lmdb.h>
+#endif
+*/
+
+#define STDIN     0
+#define STDOUT    1
+
+#define BUFSIZE 1024
+
+void usage()
+{
+    printf("Workdir is defined as %s\n", WORKDIR);
+    printf(
+    "\n"
+    "Usage: cf-keycrypt [-e public-key|-d private-key|-H hostname-or-ip] -o outfile -i infile [-h]\n"
+    "\n"
+    "Use CFEngine cryptographic keys to encrypt and decrypt files, eg. files containing\n"
+    "passwords or certificates.\n"
+    "\n"
+    "  -e       Encrypt with key.\n"
+    "  -d       Decrypt with key.\n"
+    "  -i       File to encrypt/decrypt. '-' reads from stdin.\n"
+    "  -o       File to write encrypted/decrypted contents to. '-' writes to stdout.\n"
+    "  -H       Hostname/IP to encrypt for, gets data from lastseen database\n"
+    "  -h       Print help.\n"
+    "\n"
+    "Examples:\n"
+    "  Encrypt:\n"
+    "    cf-keycrypt -e /var/cfengine/ppkeys/localhost.pub -i myplain.dat -o mycrypt.dat\n"
+    "\n"
+    "  Decrypt:\n"
+    "    cf-keycrypt -d /var/cfengine/ppkeys/localhost.priv -i mycrypt.dat -o myplain.dat\n"
+    "\n"
+    "Written and maintained by Jon Henrik Bjornstad <jonhenrik@cfengineers.net>\n"
+    "\n"
+    "Copyright (C) cfengineers.net\n"
+    "\n"
+    );
+}
+
+void *get_in_addr(struct sockaddr *sa)
+{
+    if (sa->sa_family == AF_INET)
+    {
+        return &(((struct sockaddr_in*)sa)->sin_addr);
+    }
+    return &(((struct sockaddr_in6*)sa)->sin6_addr);
+}
+
+int file_exist(char *filename)
+{
+    struct stat buffer;
+    return (stat (filename, &buffer) == 0);
+}
+
+char *get_host_pubkey(char *host)
+{
+    // char *key;                            // TODO: this was unused(!)
+    // char value[BUFSIZE];                  // TODO: this was unused(!)
+    char *buffer = (char *) malloc(BUFSIZE *sizeof(char));
+    //char *keyname = NULL;
+    char hash[CF_HOSTKEY_STRING_SIZE];
+    //char buffer[BUFSIZE];
+    char ipaddress[BUFSIZE];
+    // int ecode;                            // TODO: this was unused(!)
+    struct addrinfo *result;
+    struct addrinfo *res;
+    bool found = false;
+
+    int error = getaddrinfo(host, NULL, NULL, &result);
+    if (error != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to get IP from host (getaddrinfo: %s)",
+            gai_strerror(error));
+        return NULL;
+    }
+
+    for(res = result; res != NULL && !found; res = res->ai_next)
+    {
+        inet_ntop(res->ai_family, get_in_addr((struct sockaddr *)res->ai_addr), ipaddress, sizeof(ipaddress));
```

Comment:
Now that I think about all this whole address lookup function `get_host_pubkey()`, it has many issues in the lookups, and we have `Hostname2IPString()` for that. So better use that and save some code and risk.